src/openrct2/ride/Vehicle.cpp

```diff
@@ -9409,6 +9414,49 @@ int32_t Vehicle::UpdateTrackMotionMiniGolf(int32_t* outStation)
     return _vehicleMotionTrackFlags;
 }
 
+bool Vehicle::UpdateSpeedShift()
+{
+    auto vehicleEntry = Entry();
+    if (vehicleEntry == nullptr)
+        return false;
+    if (vehicleEntry->SpeedShift.UpperVehicle != vehicle_type)
+    {
+        if (abs(_vehicleVelocityF64E08) > (vehicleEntry->SpeedShift.UpperBound << 14))
+        {
+            vehicle_type = vehicleEntry->SpeedShift.UpperVehicle;
+            return true;
+        }
+    }
+    if (vehicleEntry->SpeedShift.LowerVehicle != vehicle_type)
+    {
+        if (abs(_vehicleVelocityF64E08) < (vehicleEntry->SpeedShift.LowerBound << 14))
+        {
+            vehicle_type = vehicleEntry->SpeedShift.LowerVehicle;
+            return true;
+        }
+    }
+    return false;
+}
+
+bool Vehicle::ForceShiftDown()
+{
+    auto vehicleEntry = Entry();
+    if (vehicleEntry == nullptr)
+        return true;
+    if (vehicleEntry->SpeedShift.LowerVehicle != vehicle_type)
+    {
+        vehicle_type = vehicleEntry->SpeedShift.LowerVehicle;
+        vehicleEntry = Entry();
+        if (vehicleEntry == nullptr)
+        {
+            // if (animation_frame >= vehicleEntry->AnimationFrames)
+            //    animation_frame = 0;
```

Comment:
This is necessary already but there's no good way to tell how many frames a ride has except with https://github.com/OpenRCT2/OpenRCT2/pull/11269